controller_manager:
  ros__parameters:
    update_rate: 100  # Hz

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

joint_state_broadcaster:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint

joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint

    command_interfaces:
      - position
    
    state_interfaces:
      - position
      - velocity
    
    # --- FIX 1: DISABLE CONSTRAINTS (Set to 0) ---
    constraints:
      goal_time: 0.0
      stopped_velocity_tolerance: 0.0
      shoulder_pan_joint: {trajectory: 0.0, goal: 0.0}
      shoulder_lift_joint: {trajectory: 0.0, goal: 0.0}
      elbow_joint: {trajectory: 0.0, goal: 0.0}
      wrist_1_joint: {trajectory: 0.0, goal: 0.0}
      wrist_2_joint: {trajectory: 0.0, goal: 0.0}
      wrist_3_joint: {trajectory: 0.0, goal: 0.0}
    
    # --- FIX 2: ADD PID GAINS (Stiffens the arm) ---
    gains:
      shoulder_pan_joint: {p: 1000.0, i: 0.0, d: 10.0}
      shoulder_lift_joint: {p: 1000.0, i: 0.0, d: 10.0}
      elbow_joint:        {p: 1000.0, i: 0.0, d: 10.0}
      wrist_1_joint:      {p: 500.0,  i: 0.0, d: 5.0}
      wrist_2_joint:      {p: 500.0,  i: 0.0, d: 5.0}
      wrist_3_joint:      {p: 500.0,  i: 0.0, d: 5.0}
